### PR TITLE
[NVFuser] Upstream push 0811

### LINKED ISSUE #82784: NVFuser produces incorrect output for broadcasted dropout
```
import torch
import functorch
from functorch.compile import memory_efficient_fusion

def fn_drop(x, y):
    return torch.nn.functional.dropout(x) * y

x=torch.ones(4, device="cuda")
y=torch.ones(4,4,device="cuda")
opt_fn=memory_efficient_fusion(fn_drop)
for _ in range(100):
    out=opt_fn(x, y)
```

prints an output with randomly dropped out values instead of dropped out columns. Same for direct ts integration, don't know if it's the same for prim-based stack. 

cc @ezyang @gchanan @zou3519 @jjsjann123, @kevinstephano, @SherlockNoMad

### PR DESCRIPTION
**Pull request by jjsjann123**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83239

Syncing nvfuser devel branch to upstream master. https://github.com/csarofeen/pytorch/

Code changes includes:

- codegen improvements:
  1. double support in expression evaluator
- bug fixes:
  1. dropout fix - rework RNG to support broadcasted dropout (Fixes #82784)
  2. expand fix - Patch expand+reduction, expand+view, rework view analysis and guard
- scheduler:
  1. manual transpose schedule example
  2. WIP transpose scheduler

Commits that's in this PR from the devel branch:

```
b7435afcd22c917713c2f41a7237bc26e1183f14 Transpose scheduler, step 1 (#1854)
8a45dbf72034684eb8e18b1835b533e90b68f184 Add an example on how to manually schedule transpose (#1889)
83dbf56a9554b2efbd5416461d938fff477b0b27 Patch dropout fix (#1898)
69d3519a532250719b1aa8341b50e067b181b42d Expand+Reduction, Expand+View support, rework View analysis and guards (#1883)
15091c488e96343bdc49e3990acbf238a3b3da51 Rework RNG to correctly support broadcasted dropout (#1888)
aafe2d048aaac596e503596a41303423619f3954 Make ExpressionEvaluator support Double (#1885)
```

RUN_TORCHBENCH: nvfuser

Differential Revision: [D38657074](https://our.internmc.facebook.com/intern/diff/D38657074)

### CODE REVIEW

**Comment by davidberard98** [comment]
@jjsjann123 I see a few build errors, can you take a look?

**Comment by jjsjann123** [comment]
> @jjsjann123 I see a few build errors, can you take a look?

Already on them~ Will update you once I got them patched

**Comment by jjsjann123** [comment]
hmm. error here doesn't seem to be caused by our update: https://github.com/pytorch/pytorch/runs/7796124160?check_suite_focus=true

I'll try my luck with rebase...

**Comment by jjsjann123** [comment]
Don't see any obvious fix in the viable/strict history. Still rebased :crossed_fingers:

**Comment by davidberard98** [comment]
@jjsjann123 
> nvcc fatal   : Unknown option '-Wno-unused-variable'

is this the failure?

**Comment by jjsjann123** [comment]
> @jjsjann123
> 
> ```
> 2022-08-11T21:58:12.3415340Z [ 98%] �[32mBuilding CUDA object test_jit/CMakeFiles/test_jit.dir/__/__/torch/csrc/jit/codegen/cuda/test/test_gpu_rng.cu.o�[0m
> 2022-08-11T21:58:12.3601782Z nvcc fatal   : Unknown option '-Wno-unused-variable'
> 2022-08-11T21:58:12.3609717Z test_jit/CMakeFiles/test_jit.dir/build.make:842: recipe for target 'test_jit/CMakeFiles/test_jit.dir/__/__/torch/csrc/jit/codegen/cuda/test/test_gpu_rng.cu.o' failed
> 2022-08-11T21:58:12.3610427Z make[2]: *** [test_jit/CMakeFiles/test_jit.dir/__/__/torch/csrc/jit/codegen/cuda/test/test_gpu_rng.cu.o] Error 1
> 2022-08-11T21:58:12.3651861Z make[2]: *** Waiting for unfinished jobs....
> ```
> 
> is this the failure?

Ha, file extension... great catch :bow:

**Comment by jjsjann123** [comment]
errr... there's actually a kernel in that file and we do want nvcc.... we need to change the build... I'm patching it now.

**Comment by jjsjann123** [comment]
Looks like it's finally green now :tada: 
cc'ing @davidberard98 for import to internal

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38657074).

**Review by davidberard98 — APPROVED** [review]
nice, I think this is looking pretty much fine internally!

**Comment by jjsjann123** [comment]
Wondering if there's any issue with landing :bow:

**Comment by davidberard98** [comment]
@jjsjann123 Internal tests are timing out it seems like, can you try rebasing and I'll re-attempt the import?

**Comment by davidberard98** [comment]
actually I can rebase it :)

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38657074).

**Comment by jjsjann123** [comment]
I see the only failing CI is the one that's complaining about LOC > 2k. Would that block the merge for us, even though we are going through your internal merge workflow?

**Comment by davidberard98** [comment]
cc @seemethere @albanD regarding 2k LOC limit from pr-sanity-checks lint. I am planning to land this via internal workflow. Can I just skip the lint failure when landing?

Context: In May we agreed that we would proceed with the following process for NVFuser PRs:
- NVFuser team does their own reviews on https://github.com/csarofeen/pytorch/. Reasons: development on main branch interrupts their development workflow.
- NVFuser team does a ~weekly push to master, i.e. this PR
- Meta engineer (me) lands the PRs via internal land process so that pytorch dev infra team does not have to handle the phabricator land process.

**Comment by albanD** [comment]
Hey @davidberard98 
Thanks for raising this up. The short answer here is that yes, this is ok to ignore this signal and land this, as long as this is done via the internal workflow and not the github first workflow.

**Comment by jjsjann123** [comment]
Wondering how internal CI went and if we are still stuck with the time-out issue from last run?

**Comment by davidberard98** [comment]
Different issues this time, still waiting for some jobs to re-run

**Comment by davidberard98** [comment]
@jjsjann123 planning to rebase this over #83751 once it lands (and in the future we should add the `skip-pr-sanity-checks` label to skip this failing test)

**Comment by jjsjann123** [comment]
> @jjsjann123 planning to rebase this over #83751 once it lands (and in the future we should add the `skip-pr-sanity-checks` label to skip this failing test)

Awesome. Thanks for keeping me updated. cc'ing @kevinstephano on skpping the ci for LOC limit

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38657074).

**Comment by jjsjann123** [comment]
hmmm. OOM error. should I be worried about that?!

**Comment by davidberard98** [comment]
@jjsjann123 it looks like it's not introduced by this PR, but I also can't repro the issue on master.. if you have torchbench set up could you try running it to see if you can repro the OOM?

**Comment by jjsjann123** [comment]
> @jjsjann123 it looks like it's not introduced by this PR, but I also can't repro the issue on master.. if you have torchbench set up could you try running it to see if you can repro the OOM?

Tried it on both A100 & V100 don't get a repro. So we should be good to go~

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38657074).
